ipsec: Add support of `ipsec-interface`, DPD options and `authby`

cathay4t opened the pull request:
Add support of these libreswan Ipsec options:
 * `ipsec-interface: 'yes'|'no'|u32`
 * `dpddelay: u64`
 * `dpdtimeout: u64`
 * `dpdaction: String`
 * `authby: String`

Integration test case included but marked as OK to fail as
NetworkManager-libreswan supporting these options is not released yet.

liangwen12year commented on rust/src/lib/ifaces/ipsec.rs:
```diff
@@ -89,10 +89,79 @@ pub struct LibreswanConfig {
+            "Invalid ipsec-interface value, should be \
+            unsigned integer or yes or no",
+        )),
+    }
```
the deserializer validation should be covered by unit tests?

cathay4t replied:
Unit test case included.

liangwen12year commented on rust/src/lib/ifaces/mod.rs:
```diff
@@ -20,42 +21,45 @@ mod ovs;
 };
-pub use dummy::DummyInterface;
-pub use ethernet::{
+pub use self::dummy::DummyInterface;
```
Use the relative path `use self::` warrants another commit ?

cathay4t replied:
Removed unrelated changes

liangwen12year commented on rust/src/lib/nispor/base_iface.rs:
```diff
@@ -26,6 +26,7 @@ fn np_iface_type_to_nmstate(
         nispor::IfaceType::Vxlan => InterfaceType::Vxlan,
         nispor::IfaceType::Ipoib => InterfaceType::InfiniBand,
         nispor::IfaceType::Tun => InterfaceType::Tun,
+        nispor::IfaceType::Other(v) if v == "xfrm" => InterfaceType::Xfrm,
```
why not support the `xfrm` interface in nispor ?

cathay4t replied:
This is urgent case, release new version of nispor take times.

qinqon replied:
Can at least add a "TODO" comment ? to move it to nispor ?

ffmancera replied:
I am creating a JIRA card to track this work :)

qinqon commented on rust/src/lib/ifaces/ipsec.rs:
```diff
@@ -89,10 +89,79 @@ pub struct LibreswanConfig {
+        skip_serializing_if = "Option::is_none",
+        rename = "ipsec-interface",
+        default,
+        deserialize_with = "parse_ipsec_iface"
```
Is not easier to make ipsec_interface a struct with a bool and a number ? instead of this multitype "yes", "no", number string ?

qinqon replied:
@cathay4t do you have the NetworkManager ref around ?

ffmancera replied:
This is mainly for multi-compatibility. When used from network-role they can specify "1", 1, true, yes.. and so on. I know it is weird but it is a problem we need to solve. Similar for some bond options in our code

ffmancera replied:
https://gitlab.gnome.org/GNOME/NetworkManager-libreswan/-/merge_requests/25

qinqon replied:
We don't have to follow exactly the same format, we can make it more "structural" like:
```yaml
use-ipsec-interfae: true|false
ipsec-interface-number: 33
```

cathay4t replied:
no. If you have better way of parsing yes|no|u32, please state.

cathay4t replied:
I don't want to invent new stuff, just follow `ipsec.conf` manpage here. The section is named as `libreswan`, so I will refrain myself on abstracting stuff.

cathay4t replied:
This is patch to NetworkManager-libreswan for this : https://gitlab.gnome.org/GNOME/NetworkManager-libreswan/-/merge_requests/25/

I have scratch build in https://people.redhat.com/fge/ipsec_nmstate/

cathay4t replied:
Only support integer or string now. no bool support.

cathay4t committed:
ipsec: Add support of `ipsec-interface`, `authby` and DPD options

Add support of these libreswan Ipsec options:
 * `ipsec-interface: 'yes'|'no'|u32`
 * `authby: String`
 * `dpddelay: u64`
 * `dpdtimeout: u64`
 * `dpdaction: String`

Integration test case included but marked as OK to fail as
NetworkManager-libreswan supporting these options is not released yet.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
LGTM
